MB-9930 migrate from golint to revive

gidjin opened the pull request:
## Description

`golint` is one of the many golang linters we use. However it is [deprecated by the owner](https://github.com/golang/go/issues/38968). Fortunately there is a drop in replacement [reviv](https://revive.run/). This PR switches to use `revive` and fixes the few issues that `revive` flagged that `golint` did not. Also found `golint` was disabled in two files but they were not needed.

## Reviewer Notes

Revive asks that we do not use `errors.New` but `fmt.Errorf` which I fixed. We already encourage using `fmt.Errorf` with `%w` in place of `errors.Wrap` so I replaced those in the files I touched. See [Errors section of the backend guide](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/golang-guide#errors)

## Setup

Run the linter

```sh
go mod tidy
golangci-lint run -v
```

Introduce an error, use `errors.New("error")` in a file, it should be flagged by the above line.

Example change that will violate `revive` linter rules
```diff
diff --git a/pkg/services/postal_codes/postal_code_validator.go b/pkg/services/postal_codes/postal_code_validator.go
index 94c8e7d48f..4be5fd9fab 100644
--- a/pkg/services/postal_codes/postal_code_validator.go
+++ b/pkg/services/postal_codes/postal_code_validator.go
@@ -1,6 +1,8 @@
 package postalcode

 import (
+       "fmt"
+       "errors"
+
        "github.com/transcom/mymove/pkg/appcontext"
        "github.com/transcom/mymove/pkg/models"
        "github.com/transcom/mymove/pkg/route"
@@ -16,7 +18,8 @@ type postalCodeValidator struct {
 func (v postalCodeValidator) ValidatePostalCode(appCtx appcontext.AppContext, postalCode string, postalCodeType services.PostalCodeType) (bool, error) {
        _, err := route.Zip5ToLatLong(postalCode)
        if err != nil {
-               return false, err
+               a := errors.New(fmt.Sprintf("error: %s", err))
+               return false, a
        }

        switch postalCodeType {
```

## Code Review Verification Steps
* [X] Request review from a member of a different team.
* [X] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-9930) for this change

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-9930">MB-9930</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2e1484bb8ab767adf0d20f55a4d52ab2b1f1f9f7
</p>

pearl-truss commented:
@gidjin how do we add the new linter? When I run the command in the description I get

`Running error: no such linter revive`

gidjin commented:
> @gidjin how do we add the new linter? When I run the command in the description I get
> 
> `Running error: no such linter revive`

I think I ran a `go mod tidy` to get it setup. I'll add it to the directions

pearl-truss commented:
> go mod tidy

hmm I still see the error after running that

> > @gidjin how do we add the new linter? When I run the command in the description I get
> > `Running error: no such linter revive`
> 
> I think I ran a `go mod tidy` to get it setup. I'll add it to the directions

Hmm I still see the same error are running `go mod tidy`. I don't get an error if I run `revive` but when I made a test commit with the error that the linter should have caught, it passed the pre-commit hook for `golangci-lint` so I think I'm still missing a set up step

pearl-truss commented:
I can wait and see if it works for anyone else, since this could be a local to me issue

gidjin commented:
I'm also happy to pair on it with you if needed. Curious that it's not working right for you. I'm using nix for setup and maybe that's a difference?

rogeruiz commented:
@pearl-truss Could you try running this:

```

go install github.com/mgechev/revive@latest

```

I took it from [the installation instructions](https://github.com/mgechev/revive). 

For non-Nix users we'll have to update it as an additional tool folks will need to install. 

For Nix users we'll have to add to it to the `default.nix`.

reggieriser reviewed (commented):
I first ran `make clean server_build client_build build_tools` and then did a `pre-commit run -a` and everything passed (isn't this also exercising `revive` via `golangci-lint`?).  I then changed the file to introduce the `errors.New` as directed, but things still passed for me surprisingly.

I then grabbed a `brew` version of `golangci-lint` (latest version) as noted in [its instructions](https://golangci-lint.run/usage/install/#macos) so I could try your command.  Things still passed for me with the `errors.New`.

Not sure why it's not failing for me when it did for you.  Maybe I'm just missing a step.  Glad to pair if you think it'd be helpful.

felipe-lee commented:
I'm having the same issue as Reggie, I can get the revive plugin to be recognized and presumably run, but not getting any linting errors when I add the `errors.New`.
Weirdly with pre-commit I had to be very specific about running the individual file:
![image](https://user-images.githubusercontent.com/35938642/139507333-f3325561-2f41-4c04-be63-be6673082377.png)
But regardless, it still passed when it shouldn't have.

gidjin commented:
🤔  I'll try and setup a time to pair next week with at least one of you to see if we can figure out why it's not working

gidjin commented:
> @pearl-truss Could you try running this:
> 
> ```
> 
> go install github.com/mgechev/revive@latest
> ```
> 
> I took it from [the installation instructions](https://github.com/mgechev/revive).
> 
> For non-Nix users we'll have to update it as an additional tool folks will need to install.
> 
> For Nix users we'll have to add to it to the `default.nix`.

I just checked and `golangci-lint` isn't installed by nix either for me. I'm going to look into that too

pearl-truss commented:
> @pearl-truss Could you try running this:
> 
> ```
> 
> go install github.com/mgechev/revive@latest
> ```
> 
> I took it from [the installation instructions](https://github.com/mgechev/revive).
> 
> For non-Nix users we'll have to update it as an additional tool folks will need to install.
> 
> For Nix users we'll have to add to it to the `default.nix`.

I still see the ` Running error: no such linter revive`

gidjin commented:
@felipe-lee @reggieriser did you two stage your change to cause an error? I recall that I had to do that before `pre-commit` would pick it up. I did see I was using `golangci-lint` from brew so I added it to the nix configuration.

```sh
mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$!⇡] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages
0 ⚡️36% ❯ git st
On branch gidjin/MB-9930_migrate_from_golint_to_revive
Your branch is ahead of 'origin/gidjin/MB-9930_migrate_from_golint_to_revive' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pkg/migrate/ListFiles.go

no changes added to commit (use "git add" and/or "git commit -a")

mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$!⇡] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages
0 ⚡️37% ❯ git add -p
diff --git a/pkg/migrate/ListFiles.go b/pkg/migrate/ListFiles.go
index fb3fdf8662..649a3111e5 100644
--- a/pkg/migrate/ListFiles.go
+++ b/pkg/migrate/ListFiles.go
@@ -1,6 +1,7 @@
 package migrate

 import (
+       "errors"
        "fmt"
        "strings"

(1/2) Stage this hunk [y,n,q,a,d,j,J,g,/,e,?]? y
@@ -39,7 +40,7 @@ func (fh *FileHelper) ListFiles(p string, s3Client s3iface.S3API) ([]string, err
                return f.Readdirnames(-1)
        } else if strings.HasPrefix(p, "s3://") {
                if s3Client == nil {
-                       return make([]string, 0), fmt.Errorf("No s3Client provided to list files at %s", p)
+                       return make([]string, 0), errors.New(fmt.Sprintf("No s3Client provided to list files at %s", p))
                }
                parts := strings.SplitN(p[len("s3://"):], "/", 2)
                if len(parts) == 2 {
(2/2) Stage this hunk [y,n,q,a,d,K,g,/,e,?]? y


mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$+⇡] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages
0 ⚡️39% ❯ git ci -m "WIP"
go version...............................................................Passed
Check JSON...........................................(no files to check)Skipped
Check for merge conflicts................................................Passed
Check Yaml...........................................(no files to check)Skipped
Detect Private Key.......................................................Passed
Trim Trailing Whitespace.................................................Passed
golangci-lint............................................................Failed
- hook id: golangci-lint
- exit code: 1

pkg/migrate/ListFiles.go:43:30: errorf: should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...) (revive)
                        return make([]string, 0), errors.New(fmt.Sprintf("No s3Client provided to list files at %s", p))
                                                  ^

markdownlint.........................................(no files to check)Skipped
Shell Syntax Check...................................(no files to check)Skipped
prettier.............................................(no files to check)Skipped
eslint...............................................(no files to check)Skipped
Swagger..............................................(no files to check)Skipped
gomod................................................(no files to check)Skipped
Generate a Docs Index................................(no files to check)Skipped
Validate CircleCI config yaml........................(no files to check)Skipped
Generate a Table of Contents in Markdown files.......(no files to check)Skipped
Run hadolint Dockerfile linter...........................................Passed
migrations manifest......................................................Passed
scripts are documented...................................................Passed
Spectral yaml linter.................................(no files to check)Skipped

mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$+⇡] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages took 10s
0 ⚡️39% ❯
```

gidjin commented:
@pearl-truss I checked and you shoudn't have to install `revive` directly. Running the linter via pre-commit, like `pre-commit run golangci-lint --all-files` should run `revive` as part of it's run. Or `golangci-lint run`. 

```sh
mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$+] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages took 24s
0 ❯ pre-commit run golangci-lint --all-files
golangci-lint............................................................Failed
- hook id: golangci-lint
- exit code: 1

pkg/migrate/ListFiles.go:43:30: errorf: should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...) (revive)
                        return make([]string, 0), errors.New(fmt.Sprintf("No s3Client provided to list files at %s", p))
                                                  ^


mymove on  gidjin/MB-9930_migrate_from_golint_to_revive [$+] via 🐹 v1.17.2 via  v14.18.1 on ☁️  (us-gov-west-1) nix-env: mymove-packages took 8s
0 ❯ golangci-lint run
pkg/migrate/ListFiles.go:43:30: errorf: should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...) (revive)
                        return make([]string, 0), errors.New(fmt.Sprintf("No s3Client provided to list files at %s", p))
                                                  ^
```

pearl-truss commented:
> pre-commit run golangci-lint --all-files

Since pulling the latest, I'm not getting the `revive not found` error anymore. But when I commit code changes that include the `errors.New()` the golang-lint passes

reggieriser commented:
> @felipe-lee @reggieriser did you two stage your change to cause an error? I recall that I had to do that before `pre-commit` would pick it up. I did see I was using `golangci-lint` from brew so I added it to the nix configuration.

I'm pretty sure I did try the file both unstaged and staged as I wondered if that might be having some effect on things.  And I also had to grab `golangci-lint` from homebrew in order to run that command explicitly (which resulted in the same thing for me doing that or a `pre-commit run -a`).

felipe-lee commented:
> > @felipe-lee @reggieriser did you two stage your change to cause an error? I recall that I had to do that before `pre-commit` would pick it up. I did see I was using `golangci-lint` from brew so I added it to the nix configuration.
> 
> I'm pretty sure I did try the file both unstaged and staged as I wondered if that might be having some effect on things. And I also had to grab `golangci-lint` from homebrew in order to run that command explicitly (which resulted in the same thing for me doing that or a `pre-commit run -a`).

Same, I get the pass whether or not I stage the change. Also I've been able to run the linters in `pre-commit` even when a file isn't staged. At least for the `go` linters when I point `pre-commit` directly at the file.

felipe-lee commented on nix/default.nix:
```diff
@@ -143,20 +143,28 @@ in buildEnv {
+      ref = "refs/heads/nixpkgs-unstable";
+      rev = "68e901c68446539165b19b227f51f7fc2b3655f9";
     }) {}).hadolint
+
+    (import (builtins.fetchGit {
+      # Descriptive name to make the store path easier to identify
+      name = "golangci-lint-1.42.1";
+      url = "https://github.com/NixOS/nixpkgs/";
+      ref = "refs/heads/nixpkgs-unstable";
+      rev = "1c804c475a6267dbe4c3f38a8f8d98067802daa0";
+    }) {}).golangci-lint
```
Thanks for adding this! Can you add the corresponding line for non-nix users? It should go in the `Brewfile.local`, you should be able to just add a line like:

```shell
brew 'golangci-lint'
```

reggieriser replied:
Is this a requirement?  I've been running our pre-commit hooks without ever having brewed this.  I brewed it for this PR, but I just uninstalled it and things ran fine without it.  I think you can just `pre-commit run golangci-lint` and achieve the same thing.  My assumption is that we get this (and the included `revive`) when we install our pre-commit hooks.

gidjin replied:
I'll try removing it from nix and uninstalling locally to see if it still works with pre-commit alone

felipe-lee replied:
I think it does work without it, but the advantage is that this gives us control over what we run against whereas our pre-commit hook is set up to run against all files (even if we pass it only a single file) so it can be helpful to have the standalone version when testing things against an individual file:

![image](https://user-images.githubusercontent.com/35938642/139725456-920b8425-3d31-41cd-8240-7e75408dd639.png)

vs 

![image](https://user-images.githubusercontent.com/35938642/139725534-05aa6916-2869-4465-b8e5-f3dcb80ccf9b.png)

Though it could also be annoying to have to keep that up to date so I'd be fine with not having it.

gidjin replied:
`pre-commit` is able to run golangci-lint without it being installed by brew or nix so I'm removing that dep from the nix config instead of adding it to the brewfile.

felipe-lee reviewed (commented):
Works as expected now. Thanks for adding the `nix` install of `golangci-lint`. Just added a comment about adding it for the non-nix setup as well, but then I think it's g2g

pearl-truss reviewed (approved):
LGTM

reggieriser reviewed (approved):
LGTM!  Ran it before making any changes and everything passed.  Made the (new) changes in the PR instructions and got the expected failure.

ahobson reviewed (approved):
🚀
